binary/private/binary_disk.f90

```diff
@@ -0,0 +1,303 @@
+         the_grid(i) = 10.0_dp**(log10(the_grid_min) + (i - 1) * (log10(the_grid_max) - log10(the_grid_min)) / (n_the - 1))
+      end do
+
+      ! only T < T_max is possible to calculate
+      T_max = pow(4.0_dp / (27.0_dp * c1**2 * c2), 1.0_dp / 9.0_dp)
+
+      do i = 1, n_the
+         T_arr(i) = 0.0_dp
+      end do
+
+      do i = 1, n_the
+         the = the_grid(i)
+         ! use bisection method
+         T_left = 0.1_dp * min(the**2 / c2, T_max)
+         f1_left = f1_the_T_fL2(the, T_left, 0.0_dp, c1, c2)
+         T_right = T_max
+         do while (abs((T_left - T_right) / T_right) > tol)
+            T = (T_left + T_right) / 2.0_dp
+            f1 = f1_the_T_fL2(the, T, 0.0_dp, c1, c2)
+            if (f1 * f1_left > 0) then
+               T_left = T
+               f1_left = f1
+            else
+               T_right = T
+            end if
+         end do
+         T_arr(i) = (T_left + T_right) / 2.0_dp
+      end do
+      ! now we have obtained numerical relation between the and T
+      do i = 1, n_the
+         logT_arr(i) = log10(T_arr(i))
+         logthe_grid(i) = log10(the_grid(i))
+      end do
+      dlogthe = logthe_grid(2) - logthe_grid(1)
+
+      ! bisection to find the numerical solution to f2(the, T, fL2=0)=0
+      the_right = 1.0_dp
+      f2_right = f2_the_T_fL2(the_right, T_the_nofL2(the_right, logthe_grid, logT_arr, n_the, dlogthe), 0.0_dp, &
+                              c3, c4, M1dot, disk_alpha, omega_K, Rd, PhiRd, GM2, PhiL1, PhiL2)
+      separation_factor = 0.95_dp
+      the_left = separation_factor * the_right
+      f2_left = f2_the_T_fL2(the_left, T_the_nofL2(the_left, logthe_grid, logT_arr, n_the, dlogthe), 0.0_dp, &
+                             c3, c4, M1dot, disk_alpha, omega_K, Rd, PhiRd, GM2, PhiL1, PhiL2)
+      do while (f2_left * f2_right > 0)
+         ! need to decrease the_left
+         the_right = the_left
+         f2_right = f2_left
+         the_left = the_left * separation_factor
+         f2_left = f2_the_T_fL2(the_left, T_the_nofL2(the_left, logthe_grid, logT_arr, n_the, dlogthe), 0.0_dp, &
+                                c3, c4, M1dot, disk_alpha, omega_K, Rd, PhiRd, GM2, PhiL1, PhiL2)
+      end do
+      ! now the solution is between the_left and the_right
+      do while (abs((the_left - the_right) / the_right) > tol)
+         the = (the_left + the_right) / 2.0_dp
+         f2 = f2_the_T_fL2(the, T_the_nofL2(the, logthe_grid, logT_arr, n_the, dlogthe), 0.0_dp, &
+                           c3, c4, M1dot, disk_alpha, omega_K, Rd, PhiRd, GM2, PhiL1, PhiL2)
+         if (f2 * f2_left > 0.0_dp) then
+            the_left = the
+            f2_left = f2
+         else
+            the_right = the
+         end if
+       end do
+      ! solution
+      the = (the_left + the_right) / 2.0_dp
+      T = T_the_nofL2(the, logthe_grid, logT_arr, n_the, dlogthe)
+      the_max = sqrt(3.0_dp / 8.0_dp * c2 * T + 1.0_dp / 4.0_dp * (PhiL2 - PhiRd) / (GM2 / Rd) - 1.0_dp / 8.0_dp)
```

Comment:
As an aside, leading coefficients to always be computed into a single constant at compile time, so e.g. `0.375_dp * ...` and `3.0_dp / 8.0_dp * ...` are truly equivalent.